File: P5/QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;
import java.util.ArrayList;

/**
 *
 * @author romanlarrosa
 */
public class Jugador implements Comparable{
    private boolean encarcelado = false;
    private String nombre = "Unknown";
    private int saldo = 7500;
    
    
    protected ArrayList<TituloPropiedad> propiedades = new ArrayList();
    protected Sorpresa cartaLibertad = null;
    protected Casilla casillaActual = null;
    
    
    
    
    //constructor
    public Jugador(String name){
        nombre = name;
    }
    
    public Jugador(Jugador otro){
        encarcelado = otro.encarcelado;
        nombre = otro.nombre;
        saldo = otro.saldo;
        
        propiedades = otro.propiedades;
        cartaLibertad = otro.cartaLibertad;
        casillaActual = otro.casillaActual;
        
    }
    
    //METODOS
    boolean cancelarHipoteca(TituloPropiedad titulo){
        boolean tengoSaldo = tengoSaldo(titulo.calcularCosteCancelar());
        boolean cancelada = false;
        if(tengoSaldo){
            modificarSaldo(-titulo.calcularCosteCancelar());
            cancelada = titulo.cancelarHipoteca();
        }
        return cancelada;
    }
    
    boolean comprarTituloPropiedad(){
        boolean comprado = false;
        int costeCompra = casillaActual.getCoste();
        if (costeCompra < saldo){
            ((Calle)casillaActual).asignarPropietario(this);
            TituloPropiedad titulo = casillaActual.getTitulo();
            propiedades.add(titulo);
            comprado = true;
            modificarSaldo(-costeCompra);
        }
       return comprado;
    }
    
    protected Especulador convertirme(int fianza){
        Especulador nuevo = new Especulador(this, fianza);
        return nuevo;
    }
    
    int cuantasCasasHotelesTengo(){
        int contador = 0;
        for(TituloPropiedad i: propiedades){
            contador = i.getNumCasas() + i.getNumHoteles();
        }
        return contador;
    }
    
    protected boolean deboIrACarcel(){
        if (tengoCartaLibertad())
            System.out.println("Felicidades, tienes la carta de libertad");
        return !tengoCartaLibertad();
    }
    
    boolean deboPagarAlquiler(){
        TituloPropiedad titulo = casillaActual.getTitulo();
        boolean esDeMiPropiedad = esDeMiPropiedad(titulo);
        boolean tienePropietario = titulo.tengoPropietario();
        boolean propietarioEncarcelado = titulo.propietarioEncarcelado();
        boolean estaHipotecada = titulo.getHipotecada();
        return (!esDeMiPropiedad & tienePropietario & !propietarioEncarcelado & !estaHipotecada);
    }
    
    Sorpresa devolverCartaLibretad(){
        Sorpresa aux = cartaLibertad;
        cartaLibertad = null;
        return aux;
    }
    
    boolean edificarCasa(TituloPropiedad titulo){
        int numCasas = titulo.getNumCasas();
        boolean edificada = false;
        if(puedoEdificarCasa(titulo)){
            int costeEdificarCasa = titulo.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarCasa);
            if(tengoSaldo){
                titulo.edificarCasa();
                modificarSaldo(-costeEdificarCasa);
                edificada = true;
            }
        }
        return edificada;
    }
    
    boolean edificarHotel(TituloPropiedad titulo){
        int numCasas = titulo.getNumCasas();
        int numHoteles = titulo.getNumHoteles();
        boolean edificado = false;
        if (numCasas == 4 && puedoEdificarHotel(titulo) ){
            int costeEdificarHotel = titulo.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarHotel);
            if(tengoSaldo){
                titulo.edificarHotel();
                modificarSaldo(-costeEdificarHotel);
                edificado = true;
            }
        }
        return edificado;
    }
    
    private void eliminarDeMisPropiedades(TituloPropiedad titulo){
       propiedades.remove(titulo);
       titulo.setPropietario(null);  
    }
    
    private boolean esDeMiPropiedad(TituloPropiedad titulo){
        boolean loEs=false;
        for(TituloPropiedad i: propiedades){
            if(i.getNombre() == titulo.getNombre()){
                loEs=true;
            }
        }
        return loEs;
    }
    
    boolean estoyEnCalleLibre(){
        throw new UnsupportedOperationException("Sin Implementar");
    }
    
    Sorpresa getCartaLibertad(){
        return cartaLibertad;
    }
    
    Casilla getCasillaActual(){
        return casillaActual;
    }

    boolean getEncarcelado() {
        return encarcelado;
    }

     public String getNombre() {
        return nombre;
    }
    
    ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }

    public int getSaldo() {
        return saldo;
    }
    
    void hipotecarPropiedad(TituloPropiedad titulo){
        int costeHipoteca = titulo.hipotecar();
        modificarSaldo(costeHipoteca);
    }
    
    void irACarcel(Casilla casilla){
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    
    
    int modificarSaldo(int cantidad){
        saldo += cantidad;
        return saldo;
    }
    
    int obtenerCapital(){
        int valor = saldo;
        for(TituloPropiedad i: propiedades){
            valor += (i.getNumCasas()+i.getNumHoteles())*i.getPrecioEdificar() + i.getPrecioCompra();
            if(i.getHipotecada())
                valor-=i.getHipotecaBase();
        }
        return valor;
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedades(boolean hipotecada){
        ArrayList<TituloPropiedad> filtro = new ArrayList<>();
        for(TituloPropiedad i: propiedades){
            if(i.getHipotecada()==hipotecada){
                filtro.add(i);
            }
        }
        
        return filtro;
    }
    
    void pagarAlquiler(){
        int costeAlquiler = ((Calle)casillaActual).pagarAlquiler();
        modificarSaldo(-costeAlquiler);    
    }
    
    protected void pagarImpuesto(){
        saldo -= casillaActual.getCoste();
        
    }
    
    void pagarLibertad(int cantidad){
        boolean tengoSaldo = tengoSaldo(cantidad); 
            if ( tengoSaldo){
                setEncarcelado(false);
                modificarSaldo(-cantidad); 
            }
    }
    
    protected boolean puedoEdificarCasa(TituloPropiedad titulo){
        if(titulo.getNumCasas() < 4 && saldo > titulo.getPrecioEdificar())
            return true;
        else
            return false;
            
    }
    
    protected boolean puedoEdificarHotel(TituloPropiedad titulo){
        if(titulo.getNumHoteles()< 4 && saldo > titulo.getPrecioEdificar())
            return true;
        else
            return false;
    }
    
    void setEncarcelado(boolean encarcelado) {
        this.encarcelado = encarcelado;
    }

    /*
    void setSaldo(int saldo) {
        this.saldo = saldo;
    }
    */

    void setCartaLibertad(Sorpresa carta) {
        this.cartaLibertad = carta;
    }

    void setCasillaActual(Casilla casilla) {
        this.casillaActual = casilla;
    }
    
    void setEncarceladoSinPoderSalir(boolean encarceladoSinPoderSalir){
        throw new UnsupportedOperationException("Sin Implementar");
    }
    
    public void setHaTiradoDado(boolean haTiradoDado){
        throw new UnsupportedOperationException("Sin Implementar");
    }
    
    boolean tengoCartaLibertad(){
        if(cartaLibertad == null)
            return false;
        else
            return true;
    }
    
    protected boolean tengoSaldo(int cantidad){
        boolean tengo = true;
        if(saldo < cantidad){
            tengo = false;
        }
        return tengo;
    }
    
    void venderPropiedad(Casilla casilla){
        TituloPropiedad titulo = casilla.getTitulo();
        eliminarDeMisPropiedades(titulo);
        int precioVenta = titulo.calcularPrecioVenta();
        modificarSaldo(precioVenta);
    }

    @Override
    public String toString() {
        String prop="";
        for(int i=0; i<propiedades.size(); i++){
            prop = prop + propiedades.get(i).toString();
        }
        return "\nJugador:" + 
                "\nEncarcelado: " + encarcelado + 
                "\nNombre: " + nombre + 
                "\nSaldo: " + saldo + 
                "\nCapital: " + this.obtenerCapital() +
                "\nPropiedades: " + prop +  //for de todas las propiedades + propiedades.toString()
                "\nCarta Libertad:" + this.tengoCartaLibertad() + 
                "\nCasilla Actual:" + casillaActual.getNumeroCasilla() + '\n';
    }

    @Override
    public int compareTo(Object o) {
       int otroCapital = ((Jugador) o).obtenerCapital();
       return otroCapital - obtenerCapital();
    }
    
    
    
}

File: P5/QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb
#encoding:utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'casilla'
require_relative 'titulo_propiedad'
module ModeloQytetet
  class Jugador
    
    attr_accessor :cartaLibertad, :casillaActual, :encarcelado, :nombre, :propiedades, :saldo
    
    def initialize(nom)
      @encarcelado = false
      @nombre = nom
      @saldo = 7500
      @propiedades = Array.new
      @cartaLibertad = nil
      @casillaActual = nil
    end
    
    def self.nuevo(nombre)
      new(nombre)
    end
    
    def self.copia(otroJugador)
      new(otroJugador.nombre)
      @encarcelado = otroJugador.encarcelado
      @saldo = otroJugador.saldo
      for i in  otroJugador.propiedades
        @propiedades << i        
      end
      @cartaLibertad = otroJugador.cartaLibertad
      @casillaActual = otroJugador.casillaActual
    end
    
    def cancelarHipoteca(tit)
      tengoSaldo = tengoSaldo(tit.calcularCosteCancelar)
      cancelada = false
      if(tengoSaldo)
        modificarSaldo(-tit.calcularCosteCancelar)
        cancelada = titulo.cancelarHipoteca
      end
      return cancelada
    end
    
    protected
    def convertirme(fianza)
      return (Especulador.new(fianza))
    end
    
    public
    def getCasillaActual()
      return @casillaActual
    end
    
    def comprarTituloPropiedad()
      comprado = false
      costeCompra=@casillaActual.coste
      if(costeCompra < @saldo)
        titulo = @casillaActual.asignarPropietario(this)
        @propiedades << titulo
        comprado = true
        modificarSaldo(-costeCompra)
      end
      return comprado
    end
    
    def cuantasCasasHotelesTengo()
     contador = 0
      for i in @propiedades
        contador += i.numCasas + i.numHoteles;
      end
      return contador
    end
    
    protected
    def deboIrACarcel()
      tengoCartaLibertad()
    end
    
    public
    def deboPagarAlquiler()
      titulo = @casillaActual.titulo
      esDeMiPropiedad = esDeMiPropiedad(titulo);
      tienePropietario = titulo.tengoPropietario();
      propietarioEncarcelado = titulo.propietarioEncarcelado();
      estaHipotecada = titulo.hipotecada
      return (!esDeMiPropiedad & tienePropietario & !propietarioEncarcelado & !estaHipotecada);
    end
    
    def devolverCartaLibertad()
      aux = @cartalibertad
      @cartalibertad = nil
      return aux
    end
    
    def edificarCasa(tit)
      numCasas = @titulo.numCasas
      edificada = false
      if (puedoEdificarCasa)
        costeEdificarCasa = @titulo.precioEdificar
        tengoSaldo = tengoSaldo(costeEdificarCasa)
        if(tengoSaldo)
          @titulo.edificarCasa
          modificarSaldo(-costeEdificarCasa)
          edificar = true
        end
      end
      return edificada
    end
    
    def edificarHotel(tit)
      numCasas= @titulo.numCasas
      numHoteles = @tiutlo.numHoteles
      edificado = false
      if(numCasas >= 4 && puedoEdificarHotel)
        costeEdificarHotel = @titulo.precioEdificar
        tengoSaldo = tengoSaldo(costeEdificarHotel)
        if(tengoSaldo)
          @titulo.edificarHotel
          modificarSaldo(-costeEdificar)
          edificado = false
        end
      end
      return edificado
    end
    
    private
    def eliminarDeMisPropiedades(tit)
      @propiedades.delete(tit)
      titulo.propietario(nil)
    end
    
    def esDeMiPropiedad(tit)
      for i in @propiedades
        if (i.nombre == tit.nombre)
          return true
        else
          return false
        end
      end
    end
    
    public
    def estoyEnCalleLibre()
      raise NotImplementedError
      #return bool
    end
    
    def hipotecarPropiedad(tit)
      costeHipoteca = titulo.hipotecar
      modificarSaldo(costeHipoteca)
     
    end
    
    def irACarcel(casilla)
      @casillaActual = casilla;
      @encarcelado = true 
    end
    
    def modificarSaldo(cant)
      @saldo += cant
      return @saldo
    end
    
    def obtenerCapital()
      valor = @saldo
      for i in @propiedades
        valor += (i.numCasas + i.numHoteles)* i.precioEdificar + i.precioCompra
        if (i.hipotecada)
          valor -= i.hipotecaBase
        end
      end
      return valor
    end
    
    def obtenerPropiedades(hipotecada)
      filtro = Array.new 
      for i in @propiedades
        if ( i.hipotecada == hipotecada)
          filtro << i
        end
      end
      return filtro
    end
    
    def pagarAlquiler()
      costeAlquiler = @casillaActual.pagarAlquiler
      modificarSaldo(-costeAlquiler)
      
    end
    
    public
    def pagarImpuesto()
      @saldo -= @casillaActual.coste
    end
    
    public
    def pagarLibertad(cant)
      tengoSaldo = tengoSaldo(cant)
      if(tengoSaldo)
        encarcelado(false)
        modificarSaldo(-cant)
      end
    end
    
    protected
    def puedoEdificarCasa(titulo)
      if ( titulo.numCasas < 4 && @saldo > titulo.precioEdificar)
        return true
      else
        return false
      end
    end
    
    def puedoEdificarHotel(titulo)
      if ( titulo.numHoteles < 4 && @saldo > titulo.precioEdificar)
        return true
      else
        return false
      end
    end
    
    public
    def setEncarceladoSinPoderSalir(bool)
      raise NotImplementedError
    end
    
    public
    def setHaTiradoDado()
      raise NotImplementedError
    end
    
    def tengoCartaLibertad()
      if(@cartaLibertad == nil)
        return false
      else
        return true
      end
    end
    
    protected
    def tengoSaldo(cant)
      if(@saldo < cant)
        return false
      else
        return true
      end
      #return bool
    end
    
    public
    def venderPropiedad(casilla)
      titulo = casilla.titulo
      eliminardemisPropiedades(titulo)
      precioVenta = titulo.calcularPrecioVenta
      modificarSaldo(precioVenta)
    end
    
    def to_s()
      str = ""
      for i in @propiedades
        str = str + i.to_s
      end
      return "\nJugador: 
      Encarcelado: #{@encarcelado}
      Nombre: #{@nombre}
      Saldo: #{@saldo}
      Capital: #{this.obtenerCapital}
      Propiedades: #{str}
      Carta Libertad: #{this.cartaLibertad}
      Casilla Actual: #{@casillaActual.to_s}"
    end
    
    def <=> (otroJugador)
      otroJugador.obtenerCapital <=> obtenerCapital
    end
  
  end
end
